lib/Dialect/ESI/ESIPasses.cpp

```diff
@@ -245,13 +245,57 @@ LogicalResult PipelineStageLowering::matchAndRewrite(
   return success();
 }
 
+namespace {
+/// Eliminate back-to-back wrap-unwraps to reduce the number of ESI channels.
+struct RemoveWrapUnwrap : public OpConversionPattern<WrapValidReady> {
+public:
+  RemoveWrapUnwrap(MLIRContext *ctxt) : OpConversionPattern(ctxt) {}
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(WrapValidReady wrap, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const final;
+};
+} // anonymous namespace
+
+LogicalResult
+RemoveWrapUnwrap::matchAndRewrite(WrapValidReady wrap,
+                                  ArrayRef<Value> stageOperands,
+                                  ConversionPatternRewriter &rewriter) const {
+  auto wrapUses = wrap.chanOutput().getUseList();
+  // This pass requires that implicit forks already be lowered by the
+  // abstract-to-physical pass. Meaning exactly one user.
+  if (!wrapUses->hasOneUse())
+    return failure();
+  UnwrapValidReady unwrap =
+      dyn_cast<UnwrapValidReady>(wrapUses->getUses().begin()->getOwner());
+  if (!unwrap)
+    return failure();
+
+  unwrap.valid().getUseList()->replaceAllUsesWith(wrap.valid());
```

Comment:
Could these be something like `rewriter.replaceOp(unwrap, {wrap.rawInput(), wrap.valid()})`?

In general, I think it may be [necessary](https://mlir.llvm.org/docs/PatternRewriter/#restrictions) to try to use the `rewriter` here to place nice with other rewrites that may come.

Reply:
Also, I think it might be best to use the `stageOperands` instead of `wrap`'s inputs directly. Other patterns may have [rewritten](https://mlir.llvm.org/docs/DialectConversion/#conversion-patterns) the operands.